src/utils/task_stats_alloc/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+            let old_bytes = bytes.fetch_sub(val, Ordering::Release);
+            // If the counter reaches zero, delete the counter. Note that we've ensured there's no
+            // zero deltas in `wrap_layout`, so there'll be no more uses of the counter.
+            if old_bytes == val {
+                // This fence is needed to prevent reordering of use of the counter and deletion of
+                // the counter. Because it is marked `Release`, the decreasing of the counter
+                // synchronizes with this `Acquire` fence. This means that use of the counter
+                // happens before decreasing the counter, which happens before this fence, which
+                // happens before the deletion of the counter.
+                fence(Ordering::Acquire);
+                unsafe { Box::from_raw_in(bytes.as_mut_ptr(), System) };
+            }
+        }
+    }
+
+    #[inline(always)]
+    pub fn val(&self) -> usize {
+        self.0
+            .as_ref()
+            .expect("bytes is invalid")
+            .load(Ordering::Relaxed)
+    }
+}
+
+task_local! {
+    pub static BYTES_ALLOCATED: TaskLocalBytesAllocated;
+}
+
+pub async fn allocation_stat<Fut, T, F>(future: Fut, interval: Duration, mut report: F) -> T
+where
+    Fut: Future<Output = T>,
+    F: FnMut(usize),
+{
+    BYTES_ALLOCATED
+        .scope(TaskLocalBytesAllocated::new(), async move {
+            // The guard has the same lifetime as the counter so that the counter will keep positive
+            // in the whole scope. When the scope exits, the guard is released, so the counter can
+            // reach zero eventually and then `drop` itself.
+            let _guard = Box::new(114514);
+            let monitor = async move {
+                let mut interval = tokio::time::interval(interval);
+                loop {
+                    interval.tick().await;
+                    BYTES_ALLOCATED.with(|bytes| report(bytes.val()));
+                }
+            };
+            let output = tokio::select! {
+                biased;
+                _ = monitor => unreachable!(),
+                output = future => output,
+            };
+            output
+        })
+        .await
+}
+
+#[inline(always)]
+fn wrap_layout(layout: Layout) -> (Layout, usize) {
+    debug_assert_ne!(layout.size(), 0, "the size of layout must be non-zero");
+
+    let (wrapped_layout, offset) = Layout::new::<TaskLocalBytesAllocated>()
+        .extend(layout)
+        .expect("wrapping layout overflow");
+    let wrapped_layout = wrapped_layout.pad_to_align();
+
+    (wrapped_layout, offset)
+}
+
+pub struct TaskLocalAlloc<A>(pub A);
+
+unsafe impl<A> GlobalAlloc for TaskLocalAlloc<A>
+where
+    A: GlobalAlloc,
+{
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let (wrapped_layout, offset) = wrap_layout(layout);
+
+        BYTES_ALLOCATED
+            .try_with(|&bytes| {
+                bytes.add_unchecked(layout.size());
```

Comment:
I think the aligned size is more accurate.

```suggestion
                bytes.add_unchecked(layout.align());
```

Reply:
I guess the `align` is not what we want? 🤔

> The minimum byte alignment for a memory block of this layout.

> `size`, when rounded up to the nearest multiple of `align`,

Reply:
My fault. Misunderstood the doc. 😇